File: 프로그래머스/unrated/178871. 달리기 경주/달리기 경주.cpp
#include <string>
#include <vector>
#include <unordered_map>

using namespace std;

vector<string> solution(vector<string> players, vector<string> callings) {
    unordered_map<string, int> ranks;
    for (int i = 0; i < players.size(); i++) {
        const string &player = players[i];
        ranks[player] = i;
    }
    
    for (const string &calling : callings) {
        int &passingRank = ranks[calling];
        const string &passedPlayer = players[passingRank - 1];
        int &passedRank = ranks[passedPlayer];
        swap(players[passingRank], players[passedRank]);
        swap(passingRank, passedRank);
    }
    return players;
}

File: 프로그래머스/1/178871. 달리기 경주/달리기 경주.py
def solution(players, callings):
    ranks = {player : rank for rank, player in enumerate(players)}
    for calling in callings:
        rank = ranks[calling]
        players[rank - 1], players[rank] = players[rank], players[rank - 1]
        ranks[players[rank]] = rank
        ranks[players[rank - 1]] = rank - 1
    return players